### PR TITLE
fix privilege escalation against inter-node communication

### PR DESCRIPTION
**Pull request by aead**
## Description

This commit fixes another privilege escalation issue
abusing the inter-node communication of distributed
servers to obtain/modify the server configuration.

The inter-node communication is authenticated using
JWT-Tokens. Further, IAM users accessing the cluster
via the web UI also get a JWT token and the browser
will add this "user" JWT token to each the request.

Now, a user can extract that JWT token an can craft
HTTP POST requests for the inter-node communication
API endpoint. Since the server accepts ANY valid
JWT token it also accepts inter-node commands from
an authenticated user such that the user can execute
arbitrary commands bypassing the IAM policy engine
and impersonate other users, change its own IAM policy
or extract the admin access/secret key.

This is fixed by only accepting "admin" JWT tokens
(tokens containing the admin access key - and therefore
were generated with the admin secret key). Consequently,
only the admin user can execute such inter-node commands.

## Motivation and Context
Fixing a security issue

## Regression
no

## How Has This Been Tested?
Manually, by sending custom JWT requests to the inter-node communication endpoint

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/storage-rest-server.go [hunk 1]
```diff
@@ -412,7 +417,7 @@ func registerStorageRESTHandlers(router *mux.Router, endpoints EndpointList) {
 		server := &storageRESTServer{storage, mustGetUUID()}
 
 		subrouter := router.PathPrefix(path.Join(storageRESTPath, endpoint.Path)).Subrouter()
-
+		fmt.Println(path.Join(storageRESTPath, endpoint.Path))
```
Comment (harshavardhana) [inline cmd/storage-rest-server.go]: Please remove the debug statement

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/storage-rest-server.go (hunk 1 above)
Comment (aead) [reply cmd/storage-rest-server.go]: Oh, sry - forgot that will remove...

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#9f721674-d818-4903-9979-cc3c92d66aa7) |
| mint-xl.sh | :x: [more...](#86da7f08-5aae-425e-9343-fa381e036f82) |
| mint-compression-dist-xl.sh | :x: [more...](#e9d536b7-f125-4dab-b461-4eac64257dd5) |
| mint-worm.sh | :x: [more...](#5ef30e61-ef5e-430a-ae34-e4c287a8a461) |
| mint-dist-xl.sh | :x: [more...](#4d29563b-bde0-404f-8888-e9f3a2d3c04e) |
<a name="86da7f08-5aae-425e-9343-fa381e036f82"></a>
### 7474-1dbe056/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31507
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4f7655d279ea:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-fdoooc4p4u3cam",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 278,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.60:31507/aws-sdk-go-test-fdoooc4p4u3cam: dial tcp 72.28.97.60:31507: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="9f721674-d818-4903-9979-cc3c92d66aa7"></a>
### 7474-1dbe056/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:30806
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3e2c890d7f1b:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 3 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-5dngwxyur9wtx5",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 2595,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:30806/aws-sdk-go-test-5dngwxyur9wtx5: dial tcp 72.28.97.57:30806: connect: no route to host",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="5ef30e61-ef5e-430a-ae34-e4c287a8a461"></a>
### 7474-1dbe056/mint-worm.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:30417
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            worm
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b5f8514539a0:/mint/log /tmp/mint-logs'

(1/1) Running worm tests ... FAILED in 3 seconds
{
  "alert": "",
  "args": {
    "bucketName": "bucket-au1fsoks0hpq2grroze7cgdluw677qxd1gl0vxg0w3mr9bjf84g82yo",
    "expiry": 60000000000,
    "objectName": "testObject"
  },
  "duration": 3073,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:30417/bucket-au1fsoks0hpq2grroze7cgdluw677qxd1gl0vxg0w3mr9bjf84g82yo: dial tcp 72.28.97.57:30417: connect: no route to host",
  "function": "PutAndDelete",
  "message": "WORM_MODE ON  - CreateBucket Failed",
  "name": "test worm mode",
  "status": "FAIL"
}

Executed 0 out of 1 tests successfully.
```
<a name="4d29563b-bde0-404f-8888-e9f3a2d3c04e"></a>
### 7474-1dbe056/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31061
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ec2defa7d4fa:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 2 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-5evccrgqusodav",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120310,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 1592048ED0DB06EB, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="e9d536b7-f125-4dab-b461-4eac64257dd5"></a>
### 7474-1dbe056/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32132
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 58c469504822:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-b2o9vrvvh1wgox",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 304,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.59:32132/aws-sdk-go-test-b2o9vrvvh1wgox: dial tcp 72.28.97.59:32132: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#f938bb82-62f3-45cd-9c9b-9b982ef1bc6b) |
| mint-compression-dist-xl.sh | :x: [more...](#9605fb0f-5b24-4310-981e-37cb99178879) |
| mint-dist-xl.sh | :x: [more...](#3edc816b-5faa-40cc-a235-cfbdf1683b3e) |
<a name="9605fb0f-5b24-4310-981e-37cb99178879"></a>
### 7474-e6c86f7/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32743
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3c6929fe16a3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 3 minutes and 13 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 19 seconds
(8/14) Running minio-dotnet tests ... done in 27 seconds
(9/14) Running minio-go tests ... done in 34 seconds
(10/14) Running minio-java tests ... FAILED in 35 seconds
{
  "name": "minio-java",
  "function": "listIncompleteUploads(String bucketName)",
  "duration": 1246,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=InternalError, message=We encountered an internal error, please try again., bucketName=minio-java-test-ivdvs6, objectName=null, resource=/minio-java-test-ivdvs6, requestId=159206811303DD01, hostId=null)\nrequest={method=GET, url=http://72.28.97.55:32743/minio-java-test-ivdvs6?max-uploads=1000&prefix=minio-java-test-fllqkg&delimiter=&uploads=, headers=Host: 72.28.97.55:32743\nUser-Agent: Minio (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20190403T171932Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20190403/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=500, headers=Accept-Ranges: bytes\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: Minio/DEVELOPMENT.2019-04-03T17-00-44Z\nVary: Origin\nX-Amz-Request-Id: 159206811303DD01\nX-Xss-Protection: 1; mode=block\nDate: Wed, 03 Apr 2019 17:19:32 GMT\nTransfer-Encoding: chunked\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1190), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executeGet(MinioClient.java:1284), io.minio.MinioClient.listIncompleteUploads(MinioClient.java:4696), io.minio.MinioClient.access$300(MinioClient.java:152), io.minio.MinioClient$6$1.populate(MinioClient.java:4546), io.minio.MinioClient$6$1.hasNext(MinioClient.java:4581), io.minio.MinioClient.removeIncompleteUpload(MinioClient.java:4930), FunctionalTest.listIncompleteUploads_test1(FunctionalTest.java:1621), FunctionalTest.runTests(FunctionalTest.java:2991), FunctionalTest.main(FunctionalTest.java:3121)]"
}

Executed 9 out of 14 tests successfully.
```
<a name="3edc816b-5faa-40cc-a235-cfbdf1683b3e"></a>
### 7474-e6c86f7/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31990
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2dc8d0cb4e3f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-zrk9iqvmc1iyzz",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 382,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.51:31990/aws-sdk-go-test-zrk9iqvmc1iyzz: dial tcp 72.28.97.51:31990: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="f938bb82-62f3-45cd-9c9b-9b982ef1bc6b"></a>
### 7474-e6c86f7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30486
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5c3b101c3558:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-3g4vmfbg21uqph",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 270,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.56:30486/aws-sdk-go-test-3g4vmfbg21uqph: dial tcp 72.28.97.56:30486: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**